File: Weather Dashboard/backend/app/schemas/weather.py
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Optional
from datetime import datetime

class CurrentWeather(BaseModel):
    temperature: float = Field(..., description="Current temperature")
    feels_like: float = Field(..., description="Feels-like temperature")
    condition: str = Field(..., description="Weather condition description")
    icon: str = Field(..., description="Weather icon code")
    humidity: float = Field(..., description="Relative humidity percentage")
    wind_speed: float = Field(..., description="Wind speed")
    wind_direction: float = Field(..., description="Wind direction in degrees")
    pressure: float = Field(..., description="Atmospheric pressure at sea level in hPa")
    visibility: Optional[float] = Field(None, description="Visibility in meters")
    cloud_cover: float = Field(..., description="Cloud cover percentage")
    uv_index: Optional[float] = Field(None, description="UV Index")

class HourlyForecastItem(BaseModel):
    time: datetime
    temperature: float
    feels_like: float
    condition: str
    icon: str
    precipitation_probability: float
    precipitation_amount: float
    wind_speed: float
    uv_index: Optional[float] = None

class HourlyForecast(BaseModel):
    hours: List[HourlyForecastItem]

class DailyForecastItem(BaseModel):
    time: datetime
    temp_max: float
    temp_min: float
    condition: str
    icon: str
    precipitation_probability: float
    precipitation_amount: float
    wind_speed: float
    uv_index: Optional[float] = None
    sunrise: datetime
    sunset: datetime

class DailyForecast(BaseModel):
    days: List[DailyForecastItem]

class LocationSearchItem(BaseModel):
    id: int
    name: str
    admin1: Optional[str] = Field(None, description="State/Region")
    country: str
    latitude: float
    longitude: float
    timezone: str

class LocationSearchResponse(BaseModel):
    results: List[LocationSearchItem]

class FavoriteCreate(BaseModel):
    name: str
    admin1: Optional[str] = None
    country: str
    latitude: float
    longitude: float
    timezone: str = "UTC"
    notes: Optional[str] = None

class FavoriteResponse(BaseModel):
    id: int
    name: str
    admin1: Optional[str] = None
    country: str
    latitude: float
    longitude: float
    timezone: str
    notes: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class AlertRuleCreate(BaseModel):
    city_name: str
    latitude: float
    longitude: float
    condition_type: str = Field(..., description="'rain', 'temp_high', 'temp_low', or 'wind'")
    threshold: float = Field(..., description="Numerical threshold e.g. 50% for rain or 35 for temp")

class AlertRuleResponse(BaseModel):
    id: int
    city_name: str
    latitude: float
    longitude: float
    condition_type: str
    threshold: float
    is_active: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class TriggeredAlert(BaseModel):
    rule_id: int
    city_name: str
    condition_type: str
    threshold: float
    current_value: float
    message: str

File: Weather Dashboard/frontend/src/types/weather.ts
export interface CurrentWeather {
  temperature: number;
  feels_like: number;
  condition: string;
  icon: string;
  humidity: number;
  wind_speed: number;
  wind_direction: number;
  pressure: number;
  visibility: number | null;
  cloud_cover: number;
  uv_index: number | null;
}

export interface HourlyForecastItem {
  time: string;
  temperature: number;
  feels_like: number;
  condition: string;
  icon: string;
  precipitation_probability: number;
  precipitation_amount: number;
  wind_speed: number;
  uv_index: number | null;
}

export interface HourlyForecast {
  hours: HourlyForecastItem[];
}

export interface DailyForecastItem {
  time: string;
  temp_max: number;
  temp_min: number;
  condition: string;
  icon: string;
  precipitation_probability: number;
  precipitation_amount: number;
  wind_speed: number;
  uv_index: number | null;
  sunrise: string;
  sunset: string;
}

export interface DailyForecast {
  days: DailyForecastItem[];
}

export interface LocationSearchItem {
  id: number;
  name: string;
  admin1: string | null;
  country: string;
  latitude: number;
  longitude: number;
  timezone: string;
}

export interface FavoriteCityItem {
  id: number;
  name: string;
  admin1?: string | null;
  country: string;
  latitude: number;
  longitude: number;
  timezone: string;
  notes?: string | null;
  created_at: string;
}

export interface AlertRuleItem {
  id: number;
  city_name: string;
  latitude: number;
  longitude: number;
  condition_type: string;
  threshold: number;
  is_active: boolean;
  created_at: string;
}

export interface TriggeredAlertItem {
  rule_id: number;
  city_name: string;
  condition_type: string;
  threshold: number;
  current_value: number;
  message: string;
}

export interface AirQualityData {
  european_aqi?: number;
  pm10?: number;
  pm2_5?: number;
  nitrogen_dioxide?: number;
  ozone?: number;
  sulphur_dioxide?: number;
}
